File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] arr) {
        int s = 0;
        int e = arr.length - 1;
        int n = arr.length;        
        int mid = 0;
        
        if(n == 0) {
            return -1;
        }
        
        if(arr[s] < arr[e] || n == 1) {
            return arr[s];
        }
        while (s <= e) {
            mid = s + (e - s) / 2;
            if (mid < e && arr[mid] > arr[mid + 1]) {
                return arr[mid + 1];
            }
            if (mid > s && arr[mid] < arr[mid - 1]) {
                return arr[mid];
            }
            if (arr[mid] <= arr[s]) {
                e = mid - 1;
            } else {
                s = mid + 1;
            }
        }
        return - 1;
    }
}

File: 0153-find-minimum-in-rotated-sorted-array/0153-find-minimum-in-rotated-sorted-array.py
class Solution:
    def findMin(self, nums: List[int]) -> int:

        s = 0
        e = len(nums) - 1
        ans = float('inf')

        while s <= e:
            mid = s + (e-s)//2
            ans = min(nums[mid], ans)
            if nums[mid] > nums[e]:
                s = mid + 1
            else:
                e = mid - 1
        
        return ans






        # while low <= high:
        #     mid = (low + high) // 2

        #     if nums[low] <= nums[mid]:
        #         ans = min(ans, nums[low])
        #         low = mid + 1
        #     else:
        #         ans = min(ans, nums[mid])
        #         high = mid - 1

        # return ans